icu4c/source/i18n/nfsubs.cpp

```diff
@@ -1306,6 +1306,9 @@ NumeratorSubstitution::doParse(const UnicodeString& text,
 
         while (workText.length() > 0 && workPos.getIndex() != 0) {
             workPos.setIndex(0);
+            if (getRuleSet() == nullptr) {
```

Comment:
Is this fix simply papering-over some other underlying problem that should have been detected and handled earlier? Is bailing out at this point safe, or are things left in some inconsistent state that could cause further problems? I think some deeper analysis of what actually happened with this fuzz test case is needed.

Reply:
How is this possible? A test would help.

The way that this is written is harder to read. Perhaps the following would be clearer.

```
            if (getRuleSet() != nullptr) {
                getRuleSet()->parse(workText, workPos, 1, nonNumericalExecutedRuleMask, recursionCount, temp); // parse zero or nothing at all
            }
            // else consider it a parse failure.
```

Reply:
done. Add java and C++ unit tests. Fix both java and c++